src/UnityExtension/Assets/Editor/GitHub.Unity/UI/TreeControl.cs

```diff
@@ -257,26 +255,116 @@ public void Blur()
             RequiresRepaint = true;
         }
 
-        private void ToggleNodeCheck(int idx, TreeNode node)
+        private void ToggleNodeChecked(int idx, TreeNode node)
         {
+            var isChecked = false;
+
+            switch (node.CheckState)
+            {
+                case CheckState.Mixed:
+                case CheckState.Empty:
+                    node.CheckState = CheckState.Checked;
+                    isChecked = true;
+                    break;
+
+                case CheckState.Checked:
+                    node.CheckState = CheckState.Empty;
+                    break;
+            }
+
             if (node.IsFolder)
             {
-                
+                ToggleChildrenChecked(idx, node, isChecked);
             }
-            else
+
+            ToggleParentFoldersChecked(idx, node, isChecked);
+        }
+
+        private void ToggleChildrenChecked(int idx, TreeNode node, bool isChecked)
+        {
+            for (var i = idx + 1; i < nodes.Count && node.Level <= nodes[i].Level; i++)
             {
-                switch (node.CheckState)
+                var childNode = nodes[i];
+                if (node.Level == childNode.Level)
                 {
-                    case CheckState.Empty:
-                        node.CheckState = CheckState.Checked;
-                        break;
+                    continue;
+                }
+
+                childNode.CheckState = isChecked ? CheckState.Checked : CheckState.Empty;
 
-                    case CheckState.Checked:
-                        node.CheckState = CheckState.Empty;
-                        break;
+                if (childNode.IsFolder)
+                {
+                    ToggleChildrenChecked(i, childNode, isChecked);
                 }
+            }
+        }
+
+        private void ToggleParentFoldersChecked(int idx, TreeNode node, bool isChecked)
+        {
+            while (true)
+            {
+                if (node.Level > 0)
```

Comment:
If node.level > 0 we are also guaranteeing that idx > 0? Otherwise i = idx - 1 could through out of bounds?

Reply:
Levels run from -1 and up. So if level > 0 then idx must be > 0.

Reply:
We are also going backwards from our current position. So we don't make any assumptions about the structure, and the for loop bounds will keep us from checking idx of -1 anyway.